src/osd/PrimaryLogPG.cc

```diff
@@ -9976,6 +9982,312 @@ struct C_Flush : public Context {
+    if (r < 0) {
+      return r;
+    }
+    if (!chunks.size()) {
+      break;
+    }
+
+    // get fingerprint 
+    for (auto p : chunks) {
+      bufferlist chunk;
+      chunk.substr_of(bl, p.first, p.second);
+      hobject_t target = get_fpoid_from_chunk(soid, chunk);
+
+      chunk_info_t chunk_info(0, p.second, target);
+      // chunks issued here are different with chunk_map newly generated
+      // because the same chunks in previous snap will not be issued
+      // So, we need two data structures; the first is the issued chunk list to track
+      // issued operations, and the second is the new chunk_map to update chunk_map after 
+      // all operations are finished
+      mop->new_chunk_map[p.first] = chunk_info;
+      // skip if the same content exits in prev snap at same offset
+      if (obc->ssc->snapset.clones.size()) {
+	ObjectContextRef cobc = get_prev_clone_obc(obc);
+	if (cobc) {
+	  object_ref_delta_t refs;
+	  object_manifest_t set_chunk;
+	  set_chunk.chunk_map[p.first] = chunk_info;
+	  set_chunk.calc_refs_to_inc_on_set(
+	    &cobc->obs.oi.manifest,
+	    nullptr,
+	    refs);
+	  if (refs.is_empty()) {
+	    dout(15) << " found same chunk " << refs << dendl;
+	    continue;
+	  } 
+	}
+      }
+	
+      // make a create_or_get_ref op
+      bufferlist t;
+      ObjectOperation obj_op;
+      cls_cas_chunk_create_or_get_ref_op get_call;
+      get_call.source = soid.get_head();
+      get_call.data = chunk;
+      ::encode(get_call, t);
+      obj_op.call("cas", "chunk_create_or_get_ref", t);
+
+      // issue create_or_get_ref_op
+      C_SetDedupChunks *fin = new C_SetDedupChunks(this, soid, get_last_peering_reset(), p.first);
+      object_locator_t oloc(target);
+      unsigned flags = CEPH_OSD_FLAG_IGNORE_CACHE | CEPH_OSD_FLAG_IGNORE_OVERLAY |
+			CEPH_OSD_FLAG_RWORDERED;
+      ceph_tid_t tid = osd->objecter->mutate(
+	  target.oid, oloc, obj_op, SnapContext(),
+	  ceph::real_clock::from_ceph_timespec(obc->obs.oi.mtime),
+	  flags, new C_OnFinisher(fin, osd->get_objecter_finisher(get_pg_shard())));
+      fin->tid = tid;
+      mop->tids[p.first] = tid;
+      mop->chunks[target] = make_pair(p.first, p.second);
+      mop->num_chunks++;
+      dout(10) << __func__ << " oid: " << soid << " tid: " << tid 
+	      << " target: " << target << " offset: " << p.first
+	      << " length: " << p.second << dendl;
+    }
+    
+    cur_off += bl.length();
+  }
+
+  if (mop->tids.size()) {
+    manifest_ops[soid] = mop;
+    manifest_ops[soid]->op = op;
+  } 
+
+  if (mop->tids.size() == 0) {
+    return 0;
+  }
+  return -EINPROGRESS;
+}
+
+int PrimaryLogPG::do_cdc(const object_info_t& oi, bufferlist& bl, vector<pair<uint64_t, uint64_t>>& chunks)
+{
+  uint64_t cur_off = 0;
+  string chunk_algo = pool.info.get_dedup_chunk_algorithm_name();
+  int64_t chunk_size = pool.info.get_dedup_cdc_chunk_size();
+  uint64_t max_window_size = static_cast<uint64_t>(pool.info.get_dedup_cdc_window_size());
```

Comment:
Is this really only meant to bound how much of the object we read at once?  This method is going to read the *whole object* into memory anyway.  The while loop in start_dedup doesn't appear to make sense, this method will always read the whole object anyway.  Moreover, there's no reason to pass in a bufferlist reference here -- the only reason you do is so that start_dedup can then then run get_fpoid_from_chunk.  Why not just do that here and return the chunk_map directly?

You need to carefully rethink how do_cdc and start_dedup work.  If you are having trouble with it, I can write a cleaner version later this week or next.